Public/Src/Engine/Cache.Plugin.CacheCore/CacheCoreArtifactContentCache.cs

```diff
@@ -369,9 +369,19 @@ public async Task<string> GetDiagnosticsAsync()
             Possible<CasHash, Failure> maybeStored = await maybeOpen.ThenAsync(
                 async cache =>
                 {
+                    Contract.Assert(content.CanSeek);
+                    int addCount = 0;
+
                     var result = await Helpers.RetryOnFailureAsync(
                         async lastAttempt =>
                         {
+                            if (addCount > 0)

```

Comment:
You might want to consider always re-setting to origin. If start at origin, this first re-set is of course unnecessary. But if you don't start at origin, then the start point will change to 0 on retries. Alternatively, you could ensure consistency by noting initial Position and always re-setting to it.